plugins/ua_network_pubsub_ethernet.c

```diff
@@ -0,0 +1,422 @@
+/* This Source Code Form is subject to the terms of the Mozilla Public
+ * License, v. 2.0. If a copy of the MPL was not distributed with this
+ * file, You can obtain one at http://mozilla.org/MPL/2.0/.
+ *
+ *   Copyright 2018 (c) Kontron Europe GmbH (Author: Rudolf Hoyler)
+ */
+
+#include "ua_plugin_network.h"
+#include "ua_log_stdout.h"
+#include "ua_util.h"
+
+/* **NOTE: Currently ONLY for LINUX** */
+#include <netpacket/packet.h>
+#include <netinet/ether.h>
+
+#include "ua_network_pubsub_ethernet.h"
+
+#define ETHERTYPE_UADP  0xB62C
+
+/* internal structure */
+#define MAX_INTERFACE_NAME 64
+#define MAX_MAC_ADDR       18
+
+/* ethernet network layer specific internal data */
+typedef struct {
+    char interfaceName[MAX_INTERFACE_NAME];
+    char destinationMac[MAX_MAC_ADDR];
+    UA_UInt16 vid;
+    UA_Byte prio;
+    u_int8_t etherSourceAddr[ETH_ALEN];
+} UA_PubSubChannelDataEth;
+
+/* for UDP/TCP exists 'UA_parseEndpointUrl' (ua_util.c).
+ * This is adapted for ethernet here locally.
+ */
+
+static UA_StatusCode
+UA_parseEndpointUrlEthernet(const UA_String *endpointUrl, UA_PubSubChannelDataEth *channelDataEth) {
+    /* related to OPC-UA specification the URL for ethernet should be:
+     *    opc.eth://<host>[:<vid>[.<pcp>]]
+     * with <host> .. MAC address, IP address or name
+     *      <vid>  .. VLAN ID
+     *      <pcp>  .. priority
+     */
+    char urlStr[64];
+    char *ptr, *ptrHost;
+    char *ptrVid = NULL;
+    char *ptrPrio = NULL;
+    int i;
+
+    if (endpointUrl->length > sizeof(urlStr)) {
+        return UA_STATUSCODE_BADINTERNALERROR;
+    }
+    sprintf(urlStr, "%.*s", (int) endpointUrl->length, endpointUrl->data);
+
+    if (strncmp(urlStr, "opc.eth://", 10) != 0) {
+        return UA_STATUSCODE_BADINTERNALERROR;
+    }
+
+    /* set host address */
+    ptrHost = urlStr + 10;
+
+    /* check for VLAN Id and priority */
+    ptr = strchr(ptrHost, ':');
+    if (ptr != NULL) {
+        *ptr = 0;
+        ptrVid = ptr + 1;
+
+        /* check for VLAN priority */
+        ptr = strchr(ptrVid, '.');
+        if (ptr != NULL) {
+            *ptr = 0;
+            ptrPrio = ptr + 1;
+        }
+    }
+    snprintf(channelDataEth->destinationMac, MAX_MAC_ADDR, "%s", ptrHost);
+    for (i = 0; channelDataEth->destinationMac[i] != 0; i++) {
+        if (channelDataEth->destinationMac[i] == ' ')
+            channelDataEth->destinationMac[i] = ':';
+    }
```

Comment:
to avoid copying, maybe we can use sscanf()? or a for loop with strtoul().

instead of storing the string representation use UA_Byte[ETH_ALEN]